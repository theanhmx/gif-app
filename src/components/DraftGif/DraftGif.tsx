import React, { ChangeEvent, FC } from 'react';

import { DraftGif } from '../../type/gif';
import Button from '../Button';

import styles from './draftGif.module.scss';

type DraftGifProps = {
	gif: DraftGif;
	onChange: (gif: DraftGif) => void;
	onRemove: () => void;
};

const DraftGifComponent: FC<DraftGifProps> = ({ gif, onChange, onRemove }) => {
	const { file, tags, source_post_url } = gif;

	const onTagsChange = (e: ChangeEvent<HTMLInputElement>) => {
		onChange({
			...gif,
			tags: e.target.value,
		});
	};

	const onSourcePostUrlChange = (e: ChangeEvent<HTMLInputElement>) => {
		onChange({
			...gif,
			source_post_url: e.target.value,
		});
	};

	const openDetail = () => {
		window.open(`/detail/${gif.id}`);
	};

	return (
		<div className={styles.draftGif}>
			<img src={URL.createObjectURL(file)} alt="gif" data-testid="gif-img" />
			<div className={styles.info}>
				{gif.id && <Button onClick={openDetail} text="Upload Success! View Now" />}
				<label htmlFor="tags" className={styles.label}>
					Tags:
				</label>
				<input
					type="text"
					name="tags"
					data-testid="tag-input"
					placeholder="pets, cat, meow"
					value={tags}
					onChange={onTagsChange}
				/>
				<label htmlFor="source_post_url" className={styles.label}>
					Source post url:
				</label>
				<input
					name="source_post_url"
					type="text"
					placeholder="http://www.mysite.com/my-post/"
					value={source_post_url}
					onChange={onSourcePostUrlChange}
					data-testid="source-input"
				/>
				<Button onClick={onRemove} text="Remove" className={styles.remove} />
			</div>
		</div>
	);
};

export default DraftGifComponent;
