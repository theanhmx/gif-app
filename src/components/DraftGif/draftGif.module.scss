.draftGif {
	display: flex;
	flex-direction: row;
	gap: 16px;
}

.info {
	display: flex;
	flex-direction: column;
	gap: 16px;
	width: 300px;
}

.label {
	color: #a6a6a6;
	font-family: interface, Helvetica Neue, helvetica, sans-serif;
	font-size: 18px;
}

.remove {
	align-self: center;
}
