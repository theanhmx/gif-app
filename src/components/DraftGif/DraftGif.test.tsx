import { fireEvent, render, screen } from '@testing-library/react';

import DraftGif from './DraftGif';

const mockGif = {
	file: new File([], 'test.gif'),
	tags: 'tag1, tag2, tag3',
	source_post_url: '',
};

describe('DraftGif', () => {
	window.URL.createObjectURL = jest.fn();

	afterEach(() => {
		(window.URL.createObjectURL as any).mockReset();
	});

	it('should render correctly', () => {
		const { container } = render(<DraftGif gif={mockGif} onChange={() => {}} onRemove={() => {}} />);
		expect(container).toMatchSnapshot();
	});

	it('should call onChange when tags are changed', () => {
		const mockOnChange = jest.fn();
		render(<DraftGif gif={mockGif} onChange={mockOnChange} onRemove={() => {}} />);

		const tagsInput = screen.getByTestId('tag-input') as HTMLInputElement;
		fireEvent.change(tagsInput, { target: { value: 'new tag' } });

		expect(mockOnChange).toHaveBeenCalledWith({
			...mockGif,
			tags: 'new tag',
		});
	});

	it('should call onChange when source input changed', () => {
		const mockOnChange = jest.fn();
		render(<DraftGif gif={mockGif} onChange={mockOnChange} onRemove={() => {}} />);

		const sourceInput = screen.getByTestId('source-input') as HTMLInputElement;
		fireEvent.change(sourceInput, { target: { value: 'new source' } });

		expect(mockOnChange).toHaveBeenCalledWith({
			...mockGif,
			source_post_url: 'new source',
		});
	});

	it('should call onRemove when remove button is clicked', () => {
		const mockOnRemove = jest.fn();
		render(<DraftGif gif={mockGif} onChange={() => {}} onRemove={mockOnRemove} />);
		const removeButton = screen.getByText('Remove');
		fireEvent.click(removeButton);
		expect(mockOnRemove).toHaveBeenCalled();
	});
});
