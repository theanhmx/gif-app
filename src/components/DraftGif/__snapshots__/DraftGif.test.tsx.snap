// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraftGif should render correctly 1`] = `
<div>
  <div
    class="draftGif"
  >
    <img
      alt="gif"
    />
    <div
      class="info"
    >
      <label
        class="label"
        for="tags"
      >
        Tags:
      </label>
      <input
        data-testid="tag-input"
        name="tags"
        placeholder="pets, cat, meow"
        type="text"
        value="tag1, tag2, tag3"
      />
      <label
        class="label"
        for="source_post_url"
      >
        Source post url:
      </label>
      <input
        data-testid="source-input"
        name="source_post_url"
        placeholder="http://www.mysite.com/my-post/"
        type="text"
        value=""
      />
      <button
        class="button remove"
      >
        Remove
      </button>
    </div>
  </div>
</div>
`;
